File: src/main/js/2021/day02.js
const fs = require("fs")
const input = fs.readFileSync(0).toString()
const data = input.split("\n").map(s => s.split(" "))

function dive(pos) {
    return data.reduce((acc, [cmd, x]) => { acc[cmd](parseInt(x)); return acc }, pos).calc()
}

function Part1Pos() {
    let horizontal = 0
    let depth = 0

    return {
        "forward": x => horizontal += x,
        "up": x => depth -= x,
        "down": x => depth += x,
        "calc": () => horizontal * depth
    }
}

function Part2Pos() {
    let horizontal = 0
    let depth = 0
    let aim = 0

    return {
        "forward": x => { horizontal += x; depth += aim * x },
        "up": x => aim -= x,
        "down": x => aim += x,
        "calc": () => horizontal * depth
    }
}

console.log(dive(Part1Pos()))
console.log(dive(Part2Pos()))

File: src/main/kotlin/aoc2021b/Day02.kt
package aoc2021b

fun main() {
    val commands = generateSequence(::readLine).map { it.split(" ") }.toList()
    dive(commands, Pos::part1)
    dive(commands, Pos::part2)
}

private inline fun dive(commands: List<List<String>>, move: Pos.(String, Int) -> Pos) = commands
    .fold(Pos()) { p, (cmd, x) -> p.move(cmd, x.toInt()) }
    .also { println(it.horizontal * it.depth) }

private data class Pos(var horizontal: Int = 0, var depth: Int = 0, var aim: Int = 0) {
    fun part1(cmd: String, x: Int): Pos {
        when (cmd) {
            "forward" -> horizontal += x
            "up" -> depth -= x
            "down" -> depth += x
        }
        return this
    }

    fun part2(cmd: String, x: Int): Pos {
        when (cmd) {
            "forward" -> {
                horizontal += x
                depth += aim * x
            }

            "up" -> aim -= x
            "down" -> aim += x
        }
        return this
    }
}